File: incoming/decompressed/evo-tactics/tools/py/validate_forms.py
#!/usr/bin/env python3
import sys, yaml, json

def load_yaml(path):
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)

def main(path="data/forms.yaml"):
    spec = load_yaml(path)
    errors, warnings = [], []

    if "global" not in spec or "starting_pi_points" not in spec["global"]:
        errors.append("missing global.starting_pi_points")
        spi = 7
    else:
        spi = int(spec["global"]["starting_pi_points"])
        if spi <= 0: errors.append("starting_pi_points must be > 0")

    pcat = spec.get("packages_catalog", [])
    pidx = {p["id"]: p for p in pcat if "id" in p}
    if not pidx:
        errors.append("packages_catalog empty or invalid")

    seen = set()
    ui = []
    for f in spec.get("forms", []):
        code = f.get("code")
        if not code:
            errors.append("form without code")
            continue
        if code in seen:
            errors.append(f"duplicate form code {code}")
        seen.add(code)

        bias = f.get("axes_bias", {})
        for k,v in bias.items():
            try:
                fv = float(v)
                if fv < -0.10 or fv > 0.10:
                    warnings.append(f"{code}: axes_bias.{k} out of range [-0.10,0.10]: {v}")
            except Exception:
                errors.append(f"{code}: axes_bias.{k} not numeric")

        pil = list(f.get("starting_pi", [])) + list(f.get("starting_pi_extra", []))
        if not pil:
            errors.append(f"{code}: no starting_pi defined")
            continue
        cost = 0
        for pid in pil:
            if pid not in pidx:
                errors.append(f"{code}: unknown package '{pid}'")
                continue
            c = int(pidx[pid].get("cost", 0))
            if c <= 0:
                errors.append(f"{code}: package '{pid}' has invalid cost {c}")
            cost += c
        if cost != spi:
            errors.append(f"{code}: PI cost sum = {cost} (must equal {spi})")

        ui.append({
            "code": code,
            "innate": f.get("innate",""),
            "pi_packages": pil,
            "pi_cost_total": cost
        })

    report = {"file": path, "errors": errors, "warnings": warnings, "forms_ui": ui}
    print(json.dumps(report, ensure_ascii=False, indent=2))
    sys.exit(0 if not errors else 1)

if __name__ == "__main__":
    path = sys.argv[1] if len(sys.argv)>1 else "data/forms.yaml"
    main(path)

File: incoming/archive_cold/devkit_scripts/2025-11-25/decompressed/evo-tactics/tools/ts/validate_forms.ts
#!/usr/bin/env node
import fs from "fs";
import yaml from "js-yaml";

function loadYaml(path) { return yaml.load(fs.readFileSync(path, "utf8")); }

function main(path = "data/forms.yaml") {
  const spec = loadYaml(path);
  const errors = [];
  const warnings = [];

  const spi = spec?.global?.starting_pi_points ?? 7;
  if (!spec?.global?.starting_pi_points) errors.push("missing global.starting_pi_points");

  const pcat = spec?.packages_catalog || [];
  const pidx = {};
  pcat.forEach((p:any) => { if (p.id) pidx[p.id] = p; });
  if (Object.keys(pidx).length === 0) errors.push("packages_catalog empty or invalid");

  const seen = new Set();
  const ui = [];
  (spec.forms || []).forEach((f:any) => {
    const code = f.code;
    if (!code) { errors.push("form without code"); return; }
    if (seen.has(code)) errors.push(`duplicate form code ${code}`);
    seen.add(code);

    const bias = f.axes_bias || {};
    Object.entries(bias).forEach(([k,v]:[string, any])=>{
      const fv = Number(v);
      if (Number.isNaN(fv)) errors.push(`${code}: axes_bias.${k} not numeric`);
      else if (fv < -0.10 || fv > 0.10) warnings.push(`${code}: axes_bias.${k} out of range [-0.10,0.10]: ${v}`);
    });

    const pil = [...(f.starting_pi||[]), ...(f.starting_pi_extra||[])];
    if (pil.length === 0) { errors.push(`${code}: no starting_pi defined`); return; }
    let cost = 0;
    pil.forEach(pid=>{
      const pkg = pidx[pid];
      if (!pkg) { errors.push(`${code}: unknown package '${pid}'`); return; }
      const c = Number(pkg.cost || 0);
      if (c <= 0) errors.push(`${code}: package '${pid}' has invalid cost ${c}`);
      cost += c;
    });
    if (cost !== spi) errors.push(`${code}: PI cost sum = ${cost} (must equal ${spi})`);

    ui.push({code, innate: f.innate || "", pi_packages: pil, pi_cost_total: cost});
  });

  const out = { file: path, errors, warnings, forms_ui: ui };
  console.log(JSON.stringify(out, null, 2));
  process.exit(errors.length ? 1 : 0);
}

const pathArg = process.argv[2] || "data/forms.yaml";
main(pathArg);
